recipe/patches/patch-ninja-exe.patch

```diff
@@ -0,0 +1,22 @@
+diff --git a/pytype/tools/analyze_project/pytype_runner.py b/pytype/tools/analyze_project/pytype_runner.py
+--- pytype/tools/analyze_project/pytype_runner.py
++++ pytype/tools/analyze_project/pytype_runner.py
+@@ -1,6 +1,7 @@
+ """Use pytype to analyze and infer types for an entire project."""
+ 
++import os
+ import collections
+ import itertools
+ import logging
+ import subprocess
+@@ -36,8 +37,10 @@
+ 
+ 
+ def _get_executable(binary, module=None):
+   """Get the path to the executable with the given name."""
++  if binary == "ninja":
++    return [os.path.join(os.environ.get("CONDA_PREFIX", None), "bin", "ninja")]
```

Comment:
What is this *actually* supposed to do when os.environ.get() falls back on None?

Reply:
If this is a patch specifically for the conda package, it *may* be OK. But ideally, we'd have a patch that the pytype folks could apply for universal use,  So relying on CONDA_PREFIX is not great.

How about something like:

```
try:
    subprocess.call('ninja')
    ninja_exe = 'ninja'
except FileNotFoundError:
    ninja_exe = 'python -m ninja'
```
or
```
try:
    import ninja
    ninja_exe = 'python -m ninja'
except ImportError:
    ninja_exe = 'ninja'
```   

which would prioritize teh pip-installed ninja if it's there.

But ideally this would get fixed upstream -- if pytype actually relies on something specific about the python ninja package, or even is only tested with it, then maybe we should jsut follow theri wishes and provide the package.

Reply:
Well, you can't join None, that's just going to print a completely bizarre:
```
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

If the environment variable ever doesn't exist.

This is why I asked what it's supposed to do. Supposed to! I'm not asking what it does, I know what it does.

I think asking "what should it fall back on then" is approaching the question from the wrong direction. The question is not what default to use to avoid a KeyError for an unset environment variable. The question instead is...

... When and why can that environment variable be unset? Is there ever a valid reason for it to be unset?

If it can't ever be unset, then don't use `.get()`, use  `os.environ['CONDA_PREFIX']`. And if that ever fails, then the user gets a much clearer error. Instead of "expected str, bytes or os.Pathlike", they get a `KeyError: 'CONDA_PREFIX'` which points the finger at conda's environment setup.

Falling back on None actually makes the error message worse.

...

Is there any specific reason to not just return `'ninja'` without a full path? My assumption is that the conda-installed `ninja` is always in $PATH whenever the conda-installed `pytype` is.

Reply:
> If this is a patch specifically for the conda package, it *may* be OK. But ideally, we'd have a patch that the pytype folks could apply for universal use, So relying on CONDA_PREFIX is not great.

Indeed.

> But ideally this would get fixed upstream -- if pytype actually relies on something specific about the python ninja package, or even is only tested with it, then maybe we should jsut follow theri wishes and provide the package.

The python ninja package is effectively identical to regular C++ ninja with the exception of a couple patches from an open PR to ninja, which add support for detecting a parent GNU make process and attach to GNU make's jobserver. This doesn't change how users interact with ninja, and doesn't change how `build.ninja` files are parsed -- it's totally invisible.

Reply:
Let's hope the pytype folks appreciate that and fix their code then.

But in any case, it would be better find a solution that is not conda-specific -- could it use "which" on *nix and "where" on Windows?

Better yet: `shutil.which`

-CHB